### PR TITLE
feat(external-router)!: rework to use canMatchFn

### LINKED ISSUE #1672: [FEAT] Using the external router should not cause two navigation events.
# :bulb: Feature request

## Feature Name
Using the external router should not cause two navigation events.

## The Desired Behavior
Navigating to a page via the external router should not cause two separate rxjs navigation actions.

## Your Use Case
Some events you might want to happen only upon a second navigation triggered by a user interaction. For example, focusing the main body of the page instead of the header could be done on the second navigation, while the initial load should focus the header. This cannot currently be done because any resolvable routes used with the external router cause two navigation events upon the first page load.

### LINKED ISSUE #1855: [FEAT] External Resolution Should Not Recurse Indefinitely
# :bulb: Feature request

## Feature Name

External Resolution Should Not Recurse Indefinitely

## The Desired Behavior

During a single external resolution navigation, the existence guard should only try to resolve and add the route a set number of times.

## Your Use Case

As an app dev, I would like to be able to configure the maximum amount of times external resolution can loop.

## Prior Work

External resolution will sometimes infinitely recurse (usually a daffodil bug) and this behavior should be caught.

### PR DESCRIPTION
**Pull request by damienwebdev**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/develop/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

Fixes: #1855 
Fixes: #1672

## What is the current behavior?

Currently, developers have to use `DaffExternalRouterExistenceGuard` which requires:

- Extraneous wildcard routing configuration 

```ts
{
    path: '**',
    canActivate: [DaffExternalRouterExistenceGuard],
    children: [],
  },
```

- Extraneous routing actions when the wildcard is reached.  If you enable your Angular routing debug log, you would see: 

```
Navigate (id:1)
NavigateCancel (id:1)
Navigate (id:2)
Navigated (id:2)
```

when navigating to any route that requires use of this guard alongside `daffPath` data.

- Complicated "insertion strategies" to inform Angular's routing configuration of the dynamic routes.

```ts
DaffExternalRouterModule.forRoot({
      failedResolutionPath: 'error',
      notFoundResolutionPath: 'not-found',
    }, [
      {
        type: 'CATEGORY',
        insertionStrategy: daffInsertDataPathStrategy,
        route: {},
      },
      {
        type: 'PRODUCT',
        insertionStrategy: daffInsertDataPathStrategy,
        route: {},
      },
    ]),
```

- Complicated dynamic route configuration

```ts
{
        matcher: daffDataPathUrlMatcher,
        data: {
          daffExternalRouteType: 'PRODUCT',
        },
        component: ProductComponent
},
```

## What is the new behavior?

The new behavior leverages the `canMatch` guards created in Angular 14. You can configure routes as follows (without any complex looking configuration):

```ts
import { ApplicationConfig } from '@angular/core';
...
import { provideExternalRouter } from '@daffodil/external-router';

export const appConfig: ApplicationConfig = {
	providers: [
		provideRouter(routes),
		provideClientHydration(),
		provideExternalRouter(),
	],
};
```

```ts
export const routes: Routes = [
	{
		path: '',
		pathMatch: 'full',
		component: HomeComponent,
	},
	{
		path: '**',
		component: TestComponent,
		canMatch: [daffExternalMatcherTypeGuard('TEST_TYPE')],
	},
	{
		path: '**',
		component: OtherTypeComponent,
		canMatch: [daffExternalMatcherTypeGuard('OTHER_TYPE')],
	},
];
```

Which is much more syntactically direct. It also fixes the original issue  as the debug log looks like:

```
Navigate (id:1)
Navigated (id:1)
```

## Does this PR introduce a breaking change?
```
[x] Yes
[ ] No
```

Users should:

1. Remove their configuration like:

```ts
DaffExternalRouterModule.forRoot({
      failedResolutionPath: 'error',
      notFoundResolutionPath: 'not-found',
    }, [
      {
        type: 'CATEGORY',
        insertionStrategy: daffInsertDataPathStrategy,
        route: {},
      },
      {
        type: 'PRODUCT',
        insertionStrategy: daffInsertDataPathStrategy,
        route: {},
      },
    ]),
```

in favor of 

```ts
provideExternalRouter({ failedResolutionPath: 'error' }),
```

2. Update `Route` configuration from:

```ts
{
        matcher: daffDataPathUrlMatcher,
        data: {
          daffExternalRouteType: 'PRODUCT',
        },
        component: ProductComponent
},
```

to:

```ts
{
    path: '**',
    component: OtherTypeComponent,
    canMatch: [daffExternalMatcherTypeGuard('PRODUCT')],
},
```

## Other information

### CODE REVIEW

File: libs/external-router/routing/src/guard/external-matcher-type.guard.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import { processErrors } from '../processors/process-errors';
+import { processRedirects } from '../processors/process-redirect';
+
+/**
+ * A `canMatch` guard function that will resolve a route externally, allowing Angular to match it.
+ *
+ * Depending on the resolution result, the route's data will contain `daffSeoData` with supplemental search engine data
+ * about the route. The data on the route is cumulative. If other data properties are set, they will be merged alongside daffSeoData.
+ */
+export const daffExternalMatcherTypeGuard = (type: string) => (route: DaffRouteWithSeoData, segments): Observable<boolean | UrlTree> => {
```
Comment (griest024) [inline libs/external-router/routing/src/guard/external-matcher-type.guard.ts]:
```suggestion
export const daffExternalMatcherTypeGuard = (type: string) => (route: DaffRouteWithSeoData, segments: Array<UrlSegment>): Observable<boolean | UrlTree> => {
```

File: libs/external-router/routing/src/guard/external-matcher-type.guard.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  const router = inject(Router);
+  const config: DaffExternalRouterConfiguration = inject(DAFF_EXTERNAL_ROUTER_CONFIG);
+  return inject(DaffExternalRouterDriver).resolve(daffConvertToPath(segments)).pipe(
+    map((resolvedRoute: DaffExternallyResolvableUrl) => processErrors(resolvedRoute)),
+    map((resolvedRoute: DaffExternallyResolvableUrl) => processRedirects(resolvedRoute)),
```
Comment (griest024) [inline libs/external-router/routing/src/guard/external-matcher-type.guard.ts]:
```suggestion
    map(processErrors),
    map(processRedirects),
```

I much prefer this functional style

File: libs/external-router/src/provide-external-router.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+} from '@angular/core';
+
+import {
+  DAFF_EXTERNAL_ROUTER_CONFIG,
+  DaffExternalRouterConfiguration,
+  daffExternalRouterConfigurationDefault,
+} from './config';
+
+export const provideExternalRouter = (config: Partial<DaffExternalRouterConfiguration> = {}): EnvironmentProviders => makeEnvironmentProviders([
+  { provide: DAFF_EXTERNAL_ROUTER_CONFIG, useValue: { ...daffExternalRouterConfigurationDefault, ...config }},
```
Comment (griest024) [inline libs/external-router/src/provide-external-router.ts]:
```suggestion
  provideDaffExternalRouterConfig(config),
```

File: libs/external-router/src/provide-external-router.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import {
+  EnvironmentProviders,
+  makeEnvironmentProviders,
+} from '@angular/core';
+
+import {
+  DAFF_EXTERNAL_ROUTER_CONFIG,
```
Comment (griest024) [inline libs/external-router/src/provide-external-router.ts]:
```suggestion
  provideDaffExternalRouterConfig,
```

File: libs/external-router/README.md [hunk 5]
```diff
@@ -1,25 +1,136 @@
...
+You can now navigate to "/test-page", "/other-page", and "/another-page"  as if it was defined in your Angular routes.
+
+## Drivers
+
+We provide a driver interface along with a few pre-fabricated drivers for you to simply drop into your app and get started with external route resolution.
+
+- [Testing](./guides/driver/testing)
+- [In-Memory](./guides/driver/in-memory)
+- [Magento](./guides/driver/magento)
+- [BYOD (Bring Your Own Driver)](./guides/driver/custom)
```
Comment (griest024) [inline libs/external-router/README.md]:
```suggestion
- [Testing](/libs/external-router/guides/driver/testing)
- [In-Memory](/libs/external-router/guides/driver/in-memory)
- [Magento](/libs/external-router/guides/driver/magento)
- [BYOD (Bring Your Own Driver)](/libs/external-router/guides/driver/custom)
```

daffio needs absolute paths to generate links properly (I want to implement support for relative soon)

File: libs/external-router/routing/src/matcher/daff-type-matcher.ts [hunk 6]
```diff
@@ -0,0 +1,6 @@
+import {
+  UrlMatcher,
+  UrlSegment,
+} from '@angular/router';
+
+export const daffExternalRouterTypeMatcher: UrlMatcher = (url: UrlSegment[]) => ({ consumed: url });
```
Comment (griest024) [inline libs/external-router/routing/src/matcher/daff-type-matcher.ts]: Document this type. In particular, address questions such as "Why does this exist?", "As an app dev, when and why would I need to use this?"

**Review by griest024 — CHANGES_REQUESTED** [review]
I love this! So much better

File: libs/external-router/routing/src/matcher/daff-type-matcher.ts (hunk 6 above)
Comment (damienwebdev) [reply libs/external-router/routing/src/matcher/daff-type-matcher.ts]: I deleted this. It was mistakenly added.

File: libs/external-router/README.md (hunk 5 above)
Comment (damienwebdev) [reply libs/external-router/README.md]: Fixed in 668ce98d0

File: libs/external-router/src/provide-external-router.ts (hunk 4 above)
Comment (damienwebdev) [reply libs/external-router/src/provide-external-router.ts]: fixed in bf605ef30

File: libs/external-router/routing/src/guard/external-matcher-type.guard.ts (hunk 2 above)
Comment (damienwebdev) [reply libs/external-router/routing/src/guard/external-matcher-type.guard.ts]: Agreed, fixed in 51703a556

File: libs/external-router/routing/src/guard/external-matcher-type.guard.ts (hunk 1 above)
Comment (damienwebdev) [reply libs/external-router/routing/src/guard/external-matcher-type.guard.ts]: Fixed in d2dd61023

File: libs/external-router/src/provide-external-router.ts (hunk 3 above)
Comment (damienwebdev) [reply libs/external-router/src/provide-external-router.ts]: fixed in https://github.com/graycoreio/daffodil/commit/bf605ef30f9314ae559074408950a7fe64d1116b

**Review by griest024 — APPROVED** [review]

**Pull request merged**